src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -111,13 +114,54 @@ private static IComponentActivator GetComponentActivatorOrDefault(IServiceProvid
         /// </summary>
         protected internal ElementReferenceContext? ElementReferenceContext { get; protected set; }
 
+        internal HotReloadContext HotReloadContext { get; } = new();
+
+        private async void RenderRootComponentsOnHotReload()
+        {
+            await Dispatcher.InvokeAsync(async () =>
+            {
+                HotReloadContext.HotReloading = true;
+
+                try
+                {
+                    _pendingTasks = new List<Task>();
+                    foreach (var componentState in _componentStateById.Values)
```

Comment:
It might work either way, but would it be viable to track a dictionary of root components instead of having the `IsRootComponent` property on `ComponentState`? This is just so the only overhead is for root components. Also would be nice to spare us from walking the entire list of components here.